Druid/Symbol.extension.st

```diff
@@ -1,5 +1,11 @@
 Extension { #name : #Symbol }
 
+{ #category : #'*Druid' }
+Symbol >> numberOfArgumentsIn: aDRPrimitiveCompilerCompiler [
+
+	^ aDRPrimitiveCompilerCompiler sourceMethod numberOfArguments
```

Comment:
Ok, so there is polymorphism, but I don't see how this work for primitive compilation, because there is no access to the pragma here!

Reply:
To understand how it works for primitives, it is best to debug from #addPrimitivesTo: (this could be triggered sending #generateIR to an DRInterpreterCompiler) . #addPrimitivesTo: walks through the #interpreterPrimitiveTable, where each primitive is specified by a symbol, e.g. #primitiveRemLargeIntegers. A CompilationUnit instance then adds a new primitive (DRPrimitiveObject) identified with this Symbol (we store this in #sourceSelector:).  When we want to get the number of arguments of this primitive specified by the symbol we do:

```smalltalk
(self compilerCompiler numberOfArgumentsFor: aString)
```

(Ideally, the PrimitiveCompiler should have no problem resolving the number of arguments of a method, whether it is a name of a primitive or a CompiledMethod, for this there is the double dispatch).

We arrive at this method which returns the number of arguments of the primitive method:

```smalltalk
CompiledMethod>>numberOfArguments

	^ (self ast pragmaNamed: #numberOfArguments:) arguments first value
```

When the primitive does not have the numberOfArguments: pragma, the method that tries to get it (#pragmaNamed:) fails here:

```smalltalk
pragmaNamed: aSymbol
	^ self pragmaNamed: aSymbol ifAbsent: [ KeyNotFound signalFor: aSymbol  ]
```

So the issue is what we do with primitive without the pragma `<numberOfArguments: >`

Reply:
For now I will implement:
```smalltalk
DRPrimitiveCompilerCompiler>>numberOfArgumentsFor: anObject
	"Answer a <Number> representing the number of arguments found in anObject (method or selector in the receiver's primitive's source)"

	^ [ anObject numberOfArgumentsIn: self ]
	on: KeyNotFound 
	do: [ : ex | ex return: 0 ]
```

Reply:
I will check Resolved and we could discuss later what we do with primitives without <numberOfArguments: >